packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -256,99 +239,65 @@ export const useGridVirtualScroller = () => {
         break;
     }
 
-    const firstRowToRender = renderContext.firstRowIndex;
-    const lastRowToRender = renderContext.lastRowIndex;
-    const firstColumnToRender = renderContext.firstColumnIndex;
-    const lastColumnToRender = renderContext.lastColumnIndex;
-
-    if (!params.rows && !currentPage.range) {
-      return [];
-    }
+    const rowModels = params.rows ?? currentPage.rows;
 
-    const renderedRows = params.rows ?? currentPage.rows.slice(firstRowToRender, lastRowToRender);
-
-    // If the selected row is not within the current range of rows being displayed,
-    // we need to render it at either the top or bottom of the rows,
-    // depending on whether it is above or below the range.
-    let isRowWithFocusedCellNotInRange = false;
-    if (
-      !isPinnedSection &&
-      indexOfRowWithFocusedCell > -1 &&
-      (firstRowToRender > indexOfRowWithFocusedCell || lastRowToRender < indexOfRowWithFocusedCell)
-    ) {
-      isRowWithFocusedCellNotInRange = true;
+    const firstRowToRender = renderContext.firstRowIndex;
+    const lastRowToRender = Math.min(renderContext.lastRowIndex, rowModels.length);
```

Comment:
Yes, in some cases `renderContext.lastRowIndex` is greater than `rowModels.length`. It's not a new bug, it's just that before this PR the `Math.min` operation was hiding as `rows.slice(firstRowToRender, lastRowToRender)`.

The root cause is that `renderContext` lags behind the rows when the "view" changes (e.g. filter modified, row expanded/collapsed, sorting modifier, etc). I've partially addressed the issue with the `useGridApiEventHandler` calls below, but some cases remain. I don't have enough time to hunt them all down in this PR.